[compiler-rt][tests] Removed the use of parentheses in compiler-rt tests with lit internal shell

Harini0924 opened the pull request:
This patch addresses compatibility issues with the lit internal shell by removing the use of subshell execution (parentheses and subshell syntax) in the `merge-posix.test` and `vptr.cpp` tests. The lit internal shell does not support parentheses, so the tests have been refactored to use separate command invocations.

This change is relevant for enabling the lit internal shell by default, as outlined in [[RFC] Enabling the Lit Internal Shell by Default](https://discourse.llvm.org/t/rfc-enabling-the-lit-internal-shell-by-default/80179)

fixes: #102401

Harini0924 committed:
[llvm-lit] Remove unsupported subshell execution in compiler-rt tests

Updated `merge-posix.test` and `vptr.cpp` to remove the use of parentheses for subshell execution in the RUN lines, ensuring compatibility with the lit internal shell.

ilovepi reviewed (approved):
Other than addressing my one comment the rest of the change seems fine to me.

Harini0924 commented on compiler-rt/test/fuzzer/merge-posix.test:
```diff
@@ -17,7 +17,8 @@ RUN: echo ....E. > %tmp/T2/5
 
 # Check that we can report an error if file size exceeded
-RUN: (ulimit -f 1; not %run %t-FullCoverageSetTest -merge=1 %tmp/T1 %tmp/T2 2>&1 | FileCheck %s --check-prefix=SIGXFSZ)
+RUN: ulimit -f 1
+RUN: not %run %t-FullCoverageSetTest -merge=1 %tmp/T1 %tmp/T2 2>&1 | FileCheck %s --check-prefix=SIGXFSZ
```
I agree with your suggestion. I'll revert the file to its original state.

Harini0924 committed:
Reverted `merge-posix.test` file back to its original state

ilovepi commented on compiler-rt/test/ubsan/TestCases/TypeCheck/vptr.cpp:
```diff
@@ -23,7 +23,9 @@
 // RUN: %env_ubsan_opts=print_stacktrace=1 %run %t dT 2>&1 | FileCheck %s --check-prefix=CHECK-DYNAMIC --allow-unused-prefixes --check-prefix=CHECK-%os-DYNAMIC --strict-whitespace
 
-// RUN: (echo "vptr_check:S"; echo "vptr_check:T"; echo "vptr_check:U") > %t.supp
+// RUN: echo "vptr_check:S" > %t.supp
+// RUN: echo "vptr_check:T" >> %t.supp
+// RUN: echo "vptr_check:U" >> %t.supp
```
Couldn't you rewrite the `echo` command w/ newlines to avoid the need for multiple RUN lines?

Alternatively, you could use use split-file instead. It will take some rewriting of the test, since you'll need to make the directory at the top of the file w/ `%t`, and then update uses of `%t` and %s`.

fmayer replied:
I think having split files just makes the test harder to read.

I don't see anything really wrong with the three `RUN` lines.

ilovepi replied:
split-file is probably overkill here, but its a pattern we use elsewhere. If we can avoid 2 subshell invocations w/o losing much (if any) readability, we probably should try to do that, IMO. It isn't a huge performance win, but I don't find it much more readable split onto 3 lines w/ file redirection than spelled out as a single `echo`, and we may as well be a little bit more efficient.

fmayer replied:
Sure, either way (except split-file) is fine with me.

Harini0924 replied:
I used `echo -e` in a single `RUN` line to create the suppression file, which simplifies the process and avoids the need for multiple subshell invocations. This keeps the script efficient and easy to maintain.  Additionally, this approach is better because it minimizes changes to the file, unlike when using `split-file`. What are your thoughts on this approach?

fmayer replied:
LGTM, but I will note that POSIX actually says `-e` is implementation dependent and that the default should be to interpret the escape characters: https://pubs.opengroup.org/onlinepubs/9799919799/. It appears that `/bin/echo` or bash's builtin do not follow this though :(

ilovepi replied:
IIRC we have other tests that use `\n` w/o `-e`, but I'm fine either way.

Harini0924 replied:
I initially tried to use `\n` without `-e`, but I encountered an error. Given that, I decided to go with `echo -e`, which worked fine in this case.

fmayer replied:
Yeah, I played around a bit and on my Linux `/bin/echo`  needed the `-e` to support the escape sequences. So did bash. zsh didn't.

POSIX is silent on that: "If the first operand consists of a '-' followed by one or more characters from the set {'e', 'E', 'n'}, or if any of the operands contain a <backslash> character, the results are implementation-defined." https://pubs.opengroup.org/onlinepubs/9799919799/utilities/echo.html

Harini0924 replied:
Thanks for the explanation—that makes sense now. Given the inconsistencies, it seems like `echo -e` is the best option. Is it okay if I stick with that?

fmayer replied:
I think so. Also, all matches for `; RUN: echo .*\\n` seem to use `-e`.

Harini0924 replied:
Yes, I just checked, and all the matches are indeed using `-e`. Thank you for the feedback and guidance.

Harini0924 committed:
Simplify suppression file creation by using a single echo command

Simplified the suppression file creation in the test script by combining multiple RUN lines into a single echo command using the -e option. This change reduces redundancy, improves readability, and makes the script more efficient by creating the suppression file with all necessary entries in one step.
